File: sdk/python/controldb/integrations/langchain.py
"""LangChain / LangGraph callback handler.

Usage::

    from controldb import ControlDB
    from controldb.integrations.langchain import ControlDBCallbackHandler

    control = ControlDB(...)
    with control.run(agent_id="aml-agent") as run:
        handler = ControlDBCallbackHandler(run)
        chain.invoke({...}, config={"callbacks": [handler]})

The handler depends on ``langchain-core``. We import lazily so the rest of
the SDK works without LangChain installed.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional


def _require_base():
    try:
        from langchain_core.callbacks import BaseCallbackHandler  # type: ignore
    except Exception as exc:  # pragma: no cover - exercised only if langchain installed
        raise ImportError(
            "langchain-core is required for ControlDBCallbackHandler. "
            "Install: pip install langchain-core"
        ) from exc
    return BaseCallbackHandler


def build_handler(run):
    BaseCallbackHandler = _require_base()

    class ControlDBCallbackHandler(BaseCallbackHandler):  # type: ignore[misc]
        def on_llm_start(self, serialized: Dict[str, Any], prompts: List[str], **kwargs: Any) -> None:
            run.emit(
                event_type="model_call.started",
                payload={"model": serialized.get("name"), "prompts": prompts},
            )

        def on_llm_end(self, response: Any, **kwargs: Any) -> None:
            run.emit(
                event_type="model_call.completed",
                payload={"response": getattr(response, "generations", str(response))},
            )

        def on_llm_error(self, error: BaseException, **kwargs: Any) -> None:
            run.emit(
                event_type="model_call.failed",
                payload={"error": {"type": type(error).__name__, "message": str(error)}},
            )

        def on_tool_start(self, serialized: Dict[str, Any], input_str: str, **kwargs: Any) -> None:
            run.emit(
                event_type="tool_call.started",
                payload={"name": serialized.get("name"), "input": input_str},
            )

        def on_tool_end(self, output: str, **kwargs: Any) -> None:
            run.emit(
                event_type="tool_call.completed",
                payload={"output": output},
            )

        def on_tool_error(self, error: BaseException, **kwargs: Any) -> None:
            run.emit(
                event_type="tool_call.failed",
                payload={"error": {"type": type(error).__name__, "message": str(error)}},
            )

    return ControlDBCallbackHandler()

File: sdk/typescript/src/integrations/langchain.ts
/**
 * LangChain callback handler integration.
 *
 * Usage:
 *   import { ControlDB } from "@controldb/sdk";
 *   import { buildHandler } from "@controldb/sdk/integrations/langchain";
 *
 *   const control = new ControlDB({ ... });
 *   await control.withRun({ agentId: "my-agent" }, async (run) => {
 *     const handler = buildHandler(run);
 *     await chain.invoke({...}, { callbacks: [handler] });
 *   });
 *
 * Requires @langchain/core to be installed.
 * Mirrors sdk/python/controldb/integrations/langchain.py exactly.
 */

import type { Run } from "../run.js";

export interface LangChainCallbackHandler {
  handleLLMStart(serialized: Record<string, unknown>, prompts: string[], kwargs?: Record<string, unknown>): Promise<void>;
  handleLLMEnd(response: unknown, kwargs?: Record<string, unknown>): Promise<void>;
  handleLLMError(error: Error, kwargs?: Record<string, unknown>): Promise<void>;
  handleToolStart(serialized: Record<string, unknown>, inputStr: string, kwargs?: Record<string, unknown>): Promise<void>;
  handleToolEnd(output: string, kwargs?: Record<string, unknown>): Promise<void>;
  handleToolError(error: Error, kwargs?: Record<string, unknown>): Promise<void>;
}

/**
 * Build a LangChain-compatible callback handler that records model and tool
 * events into the given ControlDB Run.
 *
 * The handler is a plain object implementing the LangChain BaseCallbackHandler
 * interface so it works without requiring @langchain/core at import time.
 */
export function buildHandler(run: Run): LangChainCallbackHandler {
  return {
    async handleLLMStart(
      serialized: Record<string, unknown>,
      prompts: string[]
    ): Promise<void> {
      try {
        await run.emit("model_call.started", {
          model: (serialized["name"] as string) ?? null,
          prompts,
        });
      } catch {
        // Guard: never break the LangChain pipeline
      }
    },

    async handleLLMEnd(response: unknown): Promise<void> {
      try {
        const generations =
          response !== null &&
          typeof response === "object" &&
          "generations" in (response as object)
            ? (response as Record<string, unknown>)["generations"]
            : String(response);
        await run.emit("model_call.completed", { response: generations });
      } catch {
        // Guard
      }
    },

    async handleLLMError(error: Error): Promise<void> {
      try {
        await run.emit("model_call.failed", {
          error: { type: error.name, message: error.message },
        });
      } catch {
        // Guard
      }
    },

    async handleToolStart(
      serialized: Record<string, unknown>,
      inputStr: string
    ): Promise<void> {
      try {
        await run.emit("tool_call.started", {
          name: (serialized["name"] as string) ?? null,
          input: inputStr,
        });
      } catch {
        // Guard
      }
    },

    async handleToolEnd(output: string): Promise<void> {
      try {
        await run.emit("tool_call.completed", { output });
      } catch {
        // Guard
      }
    },

    async handleToolError(error: Error): Promise<void> {
      try {
        await run.emit("tool_call.failed", {
          error: { type: error.name, message: error.message },
        });
      } catch {
        // Guard
      }
    },
  };
}
